atomic_delete: allow deletion of sstables from several prefixes

bhalevy opened the pull request:
Allow create_pending_deletion_log to delete a bunch of sstables
potentially resides in different prefixes (e.g. in the base directory
and under staging/).

The motivation arises from table::cleanup_tablet that calls compaction_group::cleanup on all cg:s via cleanup_compaction_groups.  Cleanup, in turn, calls delete_sstables_atomically on all sstables in the compaction_group, in all states, including the normal state as well as staging - hence the requirement to support deleting sstables in different sub-directories.

Also, apparently truncate calls delete_atomically for all sstables too, via table::discard_sstables, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in https://github.com/scylladb/scylladb/issues/18862 as well (although I haven't seen it yet, but I see no reason why it would happen). So the issue was apparently present since the initial implementation of the pending_delete_log. It's just that with tablet migration it is more likely to be hit.

Fixes scylladb/scylladb#18862

Needs backport to 6.0 since tablets require this capability

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10092/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10092/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10092/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2364/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10389/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10092/)
#### Build Details:
- Duration: 3 hr 18 min
- Builder: spider2.cloudius-systems.com

pwrobelse commented on sstables/sstable_directory.cc:
```diff
@@ -558,65 +558,64 @@ bool sstable_directory::compare_sstable_storage_prefix(const sstring& prefix_a,
-        } catch (...) {
-            sstlog.warn("Error while writing {}: {}. Ignoring.", pending_delete_log, std::current_exception());
+    if (ex) {
+        sstlog.warn("Error while writing {}: {}. Ignoring.", pending_delete_log, std::current_exception());
```
We are out of try-catch block. Is usage of `std::current_exception()` valid here? Shouldn't we use `ex` variable?

From stackoverflow [[1]](https://stackoverflow.com/questions/68841864/is-it-allowed-to-call-stdcurrent-exception-in-lambda-which-is-called-in-a-ca):

> An exception handler is active after its catch-clause argument is initialized and until the catch clause argument has just been destroyed. If there is not currently handled by an active handler an exception, std:current_exception will return an empty std::exception_ptr.


Toy-example which indicates, that usage of `std::current_exception` results in empty message:
https://godbolt.org/z/cvKsc4bnc

bhalevy replied:
good catch!  should `s/std::current_exception()/ex/`

bhalevy replied:
fixed in v2

avikivity commented:
What's the upgrade/downgrade impact?

bhalevy commented:
> What's the upgrade/downgrade impact?

Should have no impact. But that said I realize there's a bug in this series.
I'll add a test for it and send a fixed version, that should work also across downgrade.
Changing to draft in the meanwhile

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10113/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10113/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10113/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2385/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10409/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10113/)
:x: - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10113/)
#### Failed Tests (3/31223):
- [test_refresh_and_restart_after_compaction_strategy_change[SizeTieredCompactionStrategy-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10113/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest_with_topology_changes___test_refresh_and_restart_after_compaction_strategy_change_SizeTieredCompactionStrategy_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refresh_and_restart_after_compaction_strategy_change[SizeTieredCompactionStrategy-TimeWindowCompactionStrategy])
- [test_refresh_and_restart_after_compaction_strategy_change[SizeTieredCompactionStrategy-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10113/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest___test_refresh_and_restart_after_compaction_strategy_change_SizeTieredCompactionStrategy_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refresh_and_restart_after_compaction_strategy_change[SizeTieredCompactionStrategy-TimeWindowCompactionStrategy])
- [test_lcs_table_promotion_after_stcs_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10113/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___dtest___test_lcs_table_promotion_after_stcs_migration) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_after_stcs_migration)
#### Build Details:
- Duration: 5 hr 51 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
In v2:
- sstable_directory: use only dirlog
- sstable_directory: create_pending_deletion_log: place pending_delete log under the base directory
  - fixed formatting of pending_delete log
  - adapted unit test
  - added comments documenting why the base directory needs to be processed first
-  sstable_directory: coroutinize create_pending_deletion_log
  - on error: log saved `ex` rather than `std::current_exception()`

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10424/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2401/)
:x: - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
#### Failed Tests (2/31223):
- [test_refresh_and_restart_after_compaction_strategy_change[LeveledCompactionStrategy-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10122/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest_with_topology_changes___test_refresh_and_restart_after_compaction_strategy_change_LeveledCompactionStrategy_TimeWindowCompactionStrategy_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_refresh_and_restart_after_compaction_strategy_change[LeveledCompactionStrategy-TimeWindowCompactionStrategy])
- [test_lcs_table_promotion_after_stcs_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10122/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___dtest_with_topology_changes___test_lcs_table_promotion_after_stcs_migration) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_after_stcs_migration)
#### Build Details:
- Duration: 11 hr
- Builder: spider5.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10424/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2401/) ❌ - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10122/)
> 
> #### Failed Tests (2/31223):
> * [test_refresh_and_restart_after_compaction_strategy_change[LeveledCompactionStrategy-TimeWindowCompactionStrategy]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10122/testReport/junit/compaction_additional_test/TestCompactionAdditional/Tests___dtest_with_topology_changes___test_refresh_and_restart_after_compaction_strategy_change_LeveledCompactionStrategy_TimeWindowCompactionStrategy_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_refresh_and_restart_after_compaction_strategy_change%5BLeveledCompactionStrategy-TimeWindowCompactionStrategy%5D)
> * [test_lcs_table_promotion_after_stcs_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10122/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___dtest_with_topology_changes___test_lcs_table_promotion_after_stcs_migration) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_lcs_table_promotion_after_stcs_migration)

I can't reproduce either failures locally.
It could be that both are due to test flakiness.

bhalevy commented:
Respinned CI: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10191/

bhalevy commented:
Respinned CI: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/

avikivity commented:
> Needs backport to 6.0 since tablets require this capability

Please explain.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10469/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2444/)
:x: - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/)
#### Failed Tests (1/5690):
- [test_lcs_table_promotion_after_stcs_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10192/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___dtest_with_topology_changes___test_lcs_table_promotion_after_stcs_migration) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_lcs_table_promotion_after_stcs_migration)
#### Build Details:
- Duration: 2 hr 46 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
> > Needs backport to 6.0 since tablets require this capability
> 
> Please explain.

`table::cleanup_tablet` calls `compaction_group::cleanup` on all cg:s via `cleanup_compaction_groups`.
`cleanup`, in turn, calls `delete_sstables_atomically` on all sstables in the compaction_group, in all states, including the normal state as well as `staging` - hence the requirement to support deleting sstables in different sub-directories.

That said, I also looked at the truncate code now, and apparently truncate calls `delete_atomically` for all sstables too, via `table::discard_sstables`, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in https://github.com/scylladb/scylladb/issues/18862 as well (although I haven't seen it yet, but I see no reason why it would happen).
So the issue was apparently present since the initial implementation of the pending_delete_log.
It's just that with tablet migration it is more likely to be hit.

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10469/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2444/) ❌ - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10192/)
> 
> #### Failed Tests (1/5690):
> * [test_lcs_table_promotion_after_stcs_migration](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10192/testReport/junit/compaction_additional_test/TestLCSSSTablePromotion/Tests___dtest_with_topology_changes___test_lcs_table_promotion_after_stcs_migration) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_lcs_table_promotion_after_stcs_migration)

https://github.com/scylladb/scylladb/issues/15957

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10494/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2469/)
:x: - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/)
#### Failed Tests (1/5699):
- [test_replace_with_background_workload[use_endpoint-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10223/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_topology_changes___test_replace_with_background_workload_use_endpoint_rbo_enabled_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_background_workload[use_endpoint-rbo_enabled])
#### Build Details:
- Duration: 2 hr 38 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10494/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2469/) ❌ - [Code Coverage](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10223/)
> 
> #### Failed Tests (1/5699):
> * [test_replace_with_background_workload[use_endpoint-rbo_enabled]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10223/testReport/junit/replace_address_test/TestReplaceAddress/Tests___dtest_with_topology_changes___test_replace_with_background_workload_use_endpoint_rbo_enabled_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_replace_with_background_workload%5Buse_endpoint-rbo_enabled%5D)

https://github.com/scylladb/scylladb/issues/19645

cassandra-stress seems stuck an infinite retry loop.
completely unrelated to this change
respinning CI...

bhalevy commented:
- rebased (to get the best available baseline)

bhalevy commented:
What's going on with CI stability?!?

https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2517/testReport/repair_additional_test/TestRepairAdditional/FullDtest___full_split000___test_repair_kill_1/

https://github.com/scylladb/scylladb/issues/19223

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10280/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10280/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10280/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10542/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2517/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10280/)
#### Failed Tests (1/31282):
- [test_repair_kill_1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10280/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_1)
#### Build Details:
- Duration: 6 hr 4 min
- Builder: i-039dc36831cac0f3b (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
#### Failed Tests (1/31301):
- [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_repair_kill_3)
#### Build Details:
- Duration: 5 hr 59 min
- Builder: i-0832195b599535be0 (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
> 
> #### Failed Tests (1/31301):
> * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_repair_kill_3)

https://github.com/scylladb/scylladb/issues/19223 again :-/

I wonder if this PR somehow makes the above issue more likely, or is it just bad luck...

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10316/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10316/)
:x: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2537/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10561/)
#### Failed Tests (1/5682):
- [test_hintedhandoff_decom](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10316/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest_with_topology_changes___test_hintedhandoff_decom) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_decom)
#### Build Details:
- Duration: 3 hr 44 min
- Builder: i-02e1b0f3b2ee27b5d (m5ad.8xlarge)

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10316/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10316/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2537/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10561/)
> 
> #### Failed Tests (1/5682):
> * [test_hintedhandoff_decom](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10316/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest_with_topology_changes___test_hintedhandoff_decom) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_hintedhandoff_decom)

https://github.com/scylladb/scylla-dtest/issues/4582

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10341/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10341/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10575/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2553/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider4.cloudius-systems.com

avikivity commented:
> > #### 🔴 CI State: FAILURE
> > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
> > #### Failed Tests (1/31301):
> > 
> > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_repair_kill_3)
> 
> #19223 again :-/
> 
> I wonder if this PR somehow makes the above issue more likely, or is it just bad luck...

Do we  have an answer to this? I don't want to reduce CI stability.

bhalevy commented:
> > > #### 🔴 CI State: FAILURE
> > > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
> > > #### Failed Tests (1/31301):
> > > 
> > > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_repair_kill_3)
> > 
> > 
> > #19223 again :-/
> > I wonder if this PR somehow makes the above issue more likely, or is it just bad luck...
> 
> Do we have an answer to this? I don't want to reduce CI stability.

No.  I can't explain that.
I can run the 100 times with and without the changes here and see if the result are significantly different from each other.

bhalevy commented:
> > > > #### 🔴 CI State: FAILURE
> > > > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
> > > > #### Failed Tests (1/31301):
> > > > 
> > > > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_repair_kill_3)
> > > 
> > > 
> > > #19223 again :-/
> > > I wonder if this PR somehow makes the above issue more likely, or is it just bad luck...
> > 
> > 
> > Do we have an answer to this? I don't want to reduce CI stability.
> 
> No. I can't explain that. I can run the test 100 times with and without the changes here and see if the result are significantly different from each other.

100 times against master: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/389: 4 failures
100 times using this branch: https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/392:

bhalevy commented:
> > > > > #### 🔴 CI State: FAILURE
> > > > > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/) ❌ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2522/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10547/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10293/)
> > > > > #### Failed Tests (1/31301):
> > > > > 
> > > > > * [test_repair_kill_3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10293/testReport/junit/repair_additional_test/TestRepairAdditional/Tests___dtest_with_topology_changes___test_repair_kill_3) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+test_repair_kill_3)
> > > > 
> > > > 
> > > > #19223 again :-/
> > > > I wonder if this PR somehow makes the above issue more likely, or is it just bad luck...
> > > 
> > > 
> > > Do we have an answer to this? I don't want to reduce CI stability.
> > 
> > 
> > No. I can't explain that. I can run the test 100 times with and without the changes here and see if the result are significantly different from each other.
> 
> 100 times against master: [jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/389](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/389): 4 failures 100 times using this branch: [jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/392](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/392): 3 failures

@avikivity based on the above the excessive CI failures seen with this PR appear to be coincidental and unrelated to this change, so we can go ahead and queue this change when master is ready for business.

bhalevy commented:
@scylladb/scylla-maint please consider merging.
We need this for improving ci stability

avikivity commented:
> @scylladb/scylla-maint please consider merging. We need this for improving ci stability

How is it related to CI stability?

bhalevy commented:
> > @scylladb/scylla-maint please consider merging. We need this for improving ci stability
> 
> How is it related to CI stability?

At the moment CI doesn't run gating-dtests with tablets so it's less susceptible to https://github.com/scylladb/scylladb/issues/18862, but we want to beef-up the gating dtests suite by running them also with --tablets, and without this fix we may experience flakiness with materialized views update-generation post repair when racing with tablet migration.
The reason repair serialization with tablets migration wouldn't help, is that view update generation happens asynchronously after repair (see https://github.com/scylladb/scylladb/issues/19727)

avikivity commented:
> > > Needs backport to 6.0 since tablets require this capability
> > 
> > 
> > Please explain.
> 
> `table::cleanup_tablet` calls `compaction_group::cleanup` on all cg:s via `cleanup_compaction_groups`. `cleanup`, in turn, calls `delete_sstables_atomically` on all sstables in the compaction_group, in all states, including the normal state as well as `staging` - hence the requirement to support deleting sstables in different sub-directories.
> 
> That said, I also looked at the truncate code now, and apparently truncate calls `delete_atomically` for all sstables too, via `table::discard_sstables`, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in #18862 as well (although I haven't seen it yet, but I see no reason why it would happen). So the issue was apparently present since the initial implementation of the pending_delete_log. It's just that with tablet migration it is more likely to be hit.

It's a big series to backport.

Could we sneak in a patch that has tablet cleanup use an non-atomic deletes, or deletes atomically only sstables from a single prefix at a time, and backport only that?

bhalevy commented:
> > > > Needs backport to 6.0 since tablets require this capability
> > > 
> > > 
> > > Please explain.
> > 
> > 
> > `table::cleanup_tablet` calls `compaction_group::cleanup` on all cg:s via `cleanup_compaction_groups`. `cleanup`, in turn, calls `delete_sstables_atomically` on all sstables in the compaction_group, in all states, including the normal state as well as `staging` - hence the requirement to support deleting sstables in different sub-directories.
> > That said, I also looked at the truncate code now, and apparently truncate calls `delete_atomically` for all sstables too, via `table::discard_sstables`, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in #18862 as well (although I haven't seen it yet, but I see no reason why it would happen). So the issue was apparently present since the initial implementation of the pending_delete_log. It's just that with tablet migration it is more likely to be hit.
> 
> It's a big series to backport.
> 
> Could we sneak in a patch that has tablet cleanup use an non-atomic deletes, or deletes atomically only sstables from a single prefix at a time, and backport only that?

Can do.
The same vulnerability exists also in truncate (table::discard_sstables).
Although we haven't hit it yet, I prefer to fix it as well.

bhalevy commented:
> > > > > Needs backport to 6.0 since tablets require this capability
> > > > 
> > > > 
> > > > Please explain.
> > > 
> > > 
> > > `table::cleanup_tablet` calls `compaction_group::cleanup` on all cg:s via `cleanup_compaction_groups`. `cleanup`, in turn, calls `delete_sstables_atomically` on all sstables in the compaction_group, in all states, including the normal state as well as `staging` - hence the requirement to support deleting sstables in different sub-directories.
> > > That said, I also looked at the truncate code now, and apparently truncate calls `delete_atomically` for all sstables too, via `table::discard_sstables`, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in #18862 as well (although I haven't seen it yet, but I see no reason why it would happen). So the issue was apparently present since the initial implementation of the pending_delete_log. It's just that with tablet migration it is more likely to be hit.
> > 
> > 
> > It's a big series to backport.
> > Could we sneak in a patch that has tablet cleanup use an non-atomic deletes, or deletes atomically only sstables from a single prefix at a time, and backport only that?
> 
> Can do. The same vulnerability exists also in truncate (table::discard_sstables). Although we haven't hit it yet, I prefer to fix it as well.

@avikivity please see https://github.com/scylladb/scylladb/pull/19816

bhalevy commented:
- rebased

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10926/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10926/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10926/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3056/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10926/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11079/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
@avikivity / @xemul please re-review

bhalevy commented:
ping @avikivity please re-review

bhalevy commented:
> @avikivity / @xemul please re-review

ping @avikivity / @xemul / @scylladb/scylla-maint please re-review

bhalevy commented:
> > @avikivity / @xemul please re-review
> 
> ping @avikivity / @xemul / @scylladb/scylla-maint please re-review

Guys, what else do I need to do to advance this PR?

Cc @mykaul @dani-tweig

denesb commented on sstables/storage.cc:
```diff
@@ -517,7 +518,8 @@ class s3_storage : public sstables::storage {
+        : storage(dir)
+        , _client(std::move(client))
         , _bucket(std::move(bucket))
         , _location(std::move(dir))
```
Isn't this now duplicating `_base_dir`?

bhalevy replied:
good point, I appears you're right and we just use `_base_dir` instead of `_location`.

bhalevy replied:
fixed in v2

denesb reviewed (commented):
Does this have tests?

denesb commented:
> > > Needs backport to 6.0 since tablets require this capability
> > 
> > 
> > Please explain.
> 
> `table::cleanup_tablet` calls `compaction_group::cleanup` on all cg:s via `cleanup_compaction_groups`. `cleanup`, in turn, calls `delete_sstables_atomically` on all sstables in the compaction_group, in all states, including the normal state as well as `staging` - hence the requirement to support deleting sstables in different sub-directories.
> 
> That said, I also looked at the truncate code now, and apparently truncate calls `delete_atomically` for all sstables too, via `table::discard_sstables`, so if it happened to be executed during view update generation, i.e. when there are sstables in staging, it should hit the assertion failure reported in #18862 as well (although I haven't seen it yet, but I see no reason why it would happen). So the issue was apparently present since the initial implementation of the pending_delete_log. It's just that with tablet migration it is more likely to be hit.

This (maybe in a rephased form) should go into the cover letter. Currently it lacks any motivation as to why we need this feature and why it sould be backported.

bhalevy commented:
> Does this have tests?

@denesb, yes, there's a unit test in sstable_directory_test

bhalevy commented:
I v2 ([0cd61ca](https://github.com/scylladb/scylladb/commit/0cd61cabefc14dcd3f951520f678c7012737ef29)):
- added motivation to cover letter
- rebased
- do not duplicate _base_dir in s3_torage _location member

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/428/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3478/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11501/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/)
#### Failed Tests (1/35167):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11438/testReport/junit/raft.randomized_nemesis_test.basic_generator_test.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 3 hr 28 min
- Builder: spider7.cloudius-systems.com

bhalevy commented:
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/sstable_directory_test ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/428/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3478/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11501/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11438/)
> 
> #### Failed Tests (1/35167):
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11438/testReport/junit/raft.randomized_nemesis_test.basic_generator_test.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen+%5Bempty%5D)
> 
> #### Build Details:
> * Duration: 3 hr 28 min
> * Builder: spider7.cloudius-systems.com

Hit unrelated issue: https://github.com/scylladb/scylladb/issues/20363

Respinning

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11444/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11444/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11444/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11444/)
#### Build Details:
- Duration: 3 hr 13 min
- Builder: spider3.cloudius-systems.com

avikivity commented on replica/distributed_loader.cc:
```diff
@@ -301,6 +301,9 @@ class table_populator {
     future<> start() {
         SCYLLA_ASSERT(this_shard_id() == 0);
```
> sstable_directory: create_pending_deletion_log: place pending_delete log under the base directory

How does this work with downgrades?

1. upgrade a node
2. it crashes in a loop
3. downgrade
4. log is not detected

avikivity replied:
Ah, the base directory log will be processed, and won't notice the files have an extra component. I think it's safe.

bhalevy replied:
exactly

avikivity commented on sstables/sstable_directory.cc:
```diff
@@ -571,22 +571,22 @@ bool sstable_directory::compare_sstable_storage_prefix(const sstring& prefix_a,
-    return seastar::async([&ssts] {
-        std::unordered_map<sstring, min_max_tracker<generation_type>> gen_trackers;
-        std::unordered_map<sstring, sstring> res;
+future<sstable_directory::pending_delete_result> sstable_directory::create_pending_deletion_log(opened_directory& base_dir, const std::vector<shared_sstable>& ssts) {
```
We should encapsulate this in a class for better readability.

bhalevy replied:
encapsulate what?  the pending_delete_log functions and data structures?

avikivity commented:
Please rebase.

Does this the deletion log have tests?

bhalevy commented:
> Please rebase.

rebased

> Does this the deletion log have tests?

Yes. sstable_directory_test.test_pending_log_garbage_collection

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11481/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11481/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/sstable_directory_test
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11481/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3513/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11533/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/461/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11481/)
#### Build Details:
- Duration: 3 hr 23 min
- Builder: spider1.cloudius-systems.com

bhalevy commented:
@scylladb/scylla-maint I believe this PR is ready for merge, please consider

xemul commented on sstables/storage.hh:
```diff
@@ -74,6 +74,8 @@ public:
 class storage {
     friend class test;
 
+    std::filesystem::path _base_dir;    // Local base directory (of table)
```
Too late, maybe, but this field doesn't make sense for s3 backend. It's "location" is not location, but a key in the system.sstables, which should be schema ID, not `/var/lib/.../keyspace/table-uuid` string. And after this patch it's abused even more :(

denesb replied:
Looks like people are digging the whole you are trying to fill.

I think you should send a PR to make yourself the owner of storage.*, so you get notification when people want to change it.

denesb replied:
Ah, I see Benny did request a review from you.
